cluster mempool: add txgraph diagrams/mining/eviction

sipa opened the pull request:
Part of cluster mempool: #30289.

This adds more functionality to the txgraph module, specifically:
* `TxGraph::GetMainStagingDiagrams()`, a function to obtain feerate diagrams for both the main graph and the staged changes to it, including only the clusters that differ between the two.
* `TxGraph::GetBlockBuilder()`, a function to obtain an object which can efficiently iterate the chunks of the (main) graph from high to low chunk feerate, allowing each to be skipped or included.
* `TxGraph::GetWorstMainChunk()`, a function to obtain the last chunk that would be returned by `GetBlockBuilder()`'s returned object, intended for eviction.

sipa committed:
txgraph: Maintain chunk index (preparation)

This is preparation for exposing mining and eviction functionality in
TxGraph.

sipa committed:
txgraph: Introduce TxGraphImpl observer tracking (preparation)

This is preparation for a next commit which will introduce a class whose
objects hold references to internals in TxGraphImpl, which disallows
modifications to the graph while such objects exist.

sipa committed:
txgraph: Special-case singletons in chunk index (optimization)

sipa committed:
txgraph: Skipping end of cluster has no impact (optimization)

sipa committed:
txgraph: Add GetMainStagingDiagrams function (feature)

This allows determining whether the changes in a staging diagram unambiguously improve
the graph, through CompareChunks().

sipa committed:
txgraph: Reuse discarded chunkindex entries (optimization)

sipa committed:
txgraph: Introduce TxGraph::GetWorstMainChunk (feature)

It returns the last chunk that would be suggested for mining by BlockBuilder
objects. This is intended for eviction.

yancyribbens commented on src/cluster_linearize.h:
```diff
@@ -1336,6 +1347,38 @@ std::vector<ClusterIndex> MergeLinearizations(const DepGraph<SetType>& depgraph,
+        // Find which position to place elem in (updating j), continuously moving the elements
+        // in between forward.
+        while (place_before.Any()) {
+            // j cannot be 0 here; if it was, then there was necessarily nothing earlier which
```
```suggestion
            // j cannot be 0 or less; if it was, then there was necessarily nothing earlier which
```

ismaelsadeeq reviewed (commented):
Approach ACK will review in-depth soon

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -644,6 +647,14 @@ void Cluster::LevelDown(TxGraphImpl& graph) noexcept
     Updated(graph);
 }
 
+void Cluster::AppendChunkFeerates(std::vector<FeeFrac>& ret) const noexcept
+{
+    auto chunk_feerates = ChunkLinearization(m_depgraph, m_linearization);
+    for (const auto& feerate : chunk_feerates) {
+        ret.push_back(feerate);
+    }
+}
```
```suggestion
void Cluster::AppendChunkFeerates(std::vector<FeeFrac>& ret) const noexcept
{
    auto chunk_feerates = ChunkLinearization(m_depgraph, m_linearization);
    ret.reserve(ret.size() + chunk_feerates.size());
    ret.insert(ret.end(), chunk_feerates.begin(), chunk_feerates.end());
}
```

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -1854,6 +1865,33 @@ TxGraph::GraphIndex TxGraphImpl::CountDistinctClusters(std::span<const Ref* cons
 
+std::pair<std::vector<FeeFrac>, std::vector<FeeFrac>> TxGraphImpl::GetMainStagingDiagrams() noexcept
+{
+    Assume(m_clustersets.size() >= 2);
```
Shouldn't this be <= instead.
```suggestion
    Assume(m_clustersets.size() <= 2);

```

ismaelsadeeq commented on src/txgraph.cpp:
```diff
@@ -1854,6 +1865,33 @@ TxGraph::GraphIndex TxGraphImpl::CountDistinctClusters(std::span<const Ref* cons
+    }
+    // Do the same for the Clusters in staging themselves.
+    const auto& staging = m_clustersets.back();
+    for (int quality = 0; quality < int(QualityLevel::NONE); ++quality) {
```
why are we iterating the clusters for all quality levels for staging?

ajtowns commented on src/txgraph.h:
```diff
@@ -167,6 +169,34 @@ class TxGraph
+        /** Get the chunk that is currently suggested to be included. */
+        const std::span<Ref*>& GetCurrentChunk() noexcept { return m_current_chunk->first; }
+        /** Get the feerate of the currently suggested chunk. */
+        const FeePerWeight& GetCurrentChunkFeerate() noexcept { return m_current_chunk->second; }
```
I'm a bit surprised to see `m_current_chunk` as a persistent-ish variable that includes `Ref*` -- that makes it a bug to move `Ref`s around while you have a `BlockBuilder` in existence. I would have thought something more like:

```c++
    struct NextChunk { std::vector<Ref*> refs; FeePerWeight feerate; }
    optional<NextChunk> GetNextChunk();
    void Include(); void Skip();
    
    ...

    while (auto x = bb.GetNextChunk(); x) {
        if (x->feerate.size > remaining) { bb.Skip(); continue; }
        bb.Include();
        mine(x->refs);
    }
```

It's still a bug to move Refs around while you have a `NextChunk`, but at least that's kind-of shorter-lived than a BlockBuilder. I think that would mostly just mean moving the Next() calls into NextChunk().

ajtowns commented on src/txgraph.cpp:
```diff
@@ -289,6 +290,8 @@ class TxGraphImpl final : public TxGraph
     /** Index of ChunkData objects. */
     ChunkIndex m_chunkindex;
+    /** Number of index-observing objects in existence (BlockBuilderImpl). */
+    size_t m_chunkindex_observers{0};
```
Perhaps this stuff could be a little bit more RAII-ish? Something like:

```c++
class RefsVector : public std::vector<Ref*>
{
    TxGraph& m_graph;
public:
    template<typename... T>
    RefsView(TxGraph& graph, T&&... args) : std::vector<Ref*>{args...}, m_graph{graph} { ++m_graph.m_chunkindex_observers; }
    ~RefsView() { --m_graph.m_chunkindex_observers; }
};
```

Might be good to introduce `m_chunkindex_observers` in its own commit? Arguably having a vector of Ref's (making it invalid to do `Ref b = std::move(a);`) is a different constraint from being unable to change the linearisation at the main level (which would mess up BlockBuilder); could potentially have distinct "observer" values for each, which would allow fees to be bumped, when you just have a Ref vec, eg.

ajtowns replied:
> It's a possibility, but you mean this in addition to `BlockBuildImpl` itself also incrementing the observers count, right? Because it holds an iterator to the chunk index in GraphImpl too.

Yes. (though, technically, if `BlockBuildImpl` stores its own RefsVector it could avoid explicitly incrementing the observers count, because RefsVector does that for it)

instagibbs reviewed (commented):
just a few comments after first read through

ajtowns commented on src/txgraph.cpp:
```diff
@@ -375,6 +434,7 @@ class TxGraphImpl final : public TxGraph
     {
         auto& entry = m_entries[idx];
         Assume(entry.m_ref != nullptr);
+        Assume(m_chunkindex_observers == 0 || !entry.m_locator[0].IsPresent());
```
Should check `observers==0` in `UpdateRef` too?

ajtowns reviewed (commented):
Looks good

sipa committed:
txgraph: Introduce BlockBuilder interface (feature)

This interface lets one iterate efficiently over the chunks of the main
graph in a TxGraph, in the same order as CompareMainOrder. Each chunk
can be marked as "included" or "skipped" (and in the latter case,
dependent chunks will be skipped).

sipa committed:
txgraph: Generalize GetClusterRefs to support subsections (preparation)

This is preparation for a next commit which will need a way to extract Refs
for just individual chunks from a cluster.

sipa reviewed (commented):
Changes:
* `BlockBuilder` interface was changed, it now just has:
  * `std::optional<std::pair<std::vector<Ref*, FeePerWeight>>> GetCurrentChunk()`
  * `void Include()`
  * `void Skip()`
  Which avoids the need for it to store `Ref*` internally.
* Improved `GetMainStagingDiagrams` testing, to verify that unmodified clusters are indeed not included in the diagrams.
* Split off the introduction of the chunk index observers to a separate commit.
* Split off the generalization of `GetClusterRefs` to support subsections of a Cluster to a separate commit.

instagibbs commented on src/txgraph.cpp:
```diff
@@ -1675,6 +1677,8 @@ void Cluster::GetClusterRefs(TxGraphImpl& graph, std::span<TxGraph::Ref*> range,
         ref = entry.m_ref;
     }
+    // Return whether this was the end of the Cluster.
+    return start_pos == m_linearization.size();
```
nit: fwiw forgot start_pos was actually end_pos by this point, maybe this function should have cur_pos?

instagibbs replied:
feel free to resolve

instagibbs commented on src/txgraph.cpp:
```diff
@@ -148,8 +148,9 @@ class Cluster
     void GetDescendantRefs(const TxGraphImpl& graph, std::span<std::pair<Cluster*, DepGraphIndex>>& args, std::vector<TxGraph::Ref*>& output) noexcept;
-    /** Get a vector of Refs for all elements of this Cluster, in linearization order. */
-    void GetClusterRefs(TxGraphImpl& graph, std::span<TxGraph::Ref*> range, LinearizationIndex start_pos) noexcept;
+    /** Get a vector of Refs for all elements of this Cluster, in linearization order. Returns
```
re: commit message for a0e19959e85b2ce1a90902ab1e3fe274f9d63a1d

Might just stress that this makes `m_excluded_cluster` a lot smaller since this takes the size from O(number cluster) to O(number non-singleton clusters).

instagibbs commented on src/txgraph.cpp:
```diff
@@ -2322,13 +2328,22 @@ std::optional<std::pair<std::vector<TxGraph::Ref*>, FeePerWeight>> BlockBuilderI
+            auto start_pos = chunk_end_entry.m_main_lin_index + 1 - chunk_data.m_chunk_count;
+            bool is_end = cluster->GetClusterRefs(*m_graph, ret->first, start_pos);
+            if (is_end) {
+                m_cur_cluster = nullptr;
```
could `Assume()` here that `chunk_data.m_chunk_count > 1`, since it would have been given the special value otherwise?

instagibbs reviewed (commented):
Looking good 0630995fee22990402771547be1480b8706c76ce

No real show-stoppers except for the diagram check fuzz test conceptual issue I'm still having.

instagibbs commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -649,6 +714,50 @@ FUZZ_TARGET(txgraph)
+            assert(stage_cmp_diagram.size() + missing_stage_cmp.size() == stage_diagram.size());
+            // The missing chunks must be equal across main & staging (otherwise they couldn't have
+            // been omitted).
+            assert(missing_main_cmp == missing_stage_cmp);
```
ah! please leave a note as to that, I always forget that's possible

instagibbs commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -279,6 +290,39 @@ FUZZ_TARGET(txgraph)
+        // that GetCluster on each element of a cluster reports the cluster transactions in the same
+        // order.
+        assert(num_tx == sim.GetTransactionCount());
+        // Sort by feerate (we don't care about respecting ordering within clusters, as these are
```
Oh, I misread this, but I'm still having trouble understanding what is intended. "Sort by feerate" ... "as these are just feerates".

ajtowns commented on src/txgraph.h:
```diff
@@ -162,6 +162,10 @@ class TxGraph
+    /** Get feerate diagrams for both main and staging (which must both exist and not be
+     *  oversized), ignoring unmodified components in both. Use FeeFrac rather than FeePerWeight
+     *  so CompareChunks is usable without type-conversion. */
+    virtual std::pair<std::vector<FeeFrac>, std::vector<FeeFrac>> GetMainStagingDiagrams() noexcept = 0;
```
Yeah, guess that pushes it too far into the "not worth it"

ajtowns commented on src/txgraph.cpp:
```diff
@@ -1901,6 +1911,32 @@ TxGraph::GraphIndex TxGraphImpl::CountDistinctClusters(std::span<const Ref* cons
+    }
+    // Sort both by decreasing feerate to obtain diagrams, and return them.
+    std::sort(main_feerates.begin(), main_feerates.end(), [](auto& a, auto& b) { return a >> b; });
+    std::sort(staging_feerates.begin(), staging_feerates.end(), [](auto& a, auto& b) { return a >> b; });
```
Hmm. I didn't understand how BlockBuilder was implemented when I wrote that; BlockBuilder maintains an ordered tree of every main chunk, so is O(num_chunks) in size/performance, which is fine. What I had in mind conceptually about cluster mempool is that you'd have a structure that was O(num_clusters) that you could query for the the best/worst chunk. That would be O(n log(n)) setup, then O(log(n)) for the equivalent of a "Next()" call, for n=num_clusters. But that's not superior to how BlockBuilder works (and is probably inferior), so, doesn't sound like there's anything worth doing here.

ajtowns replied:
Thinking some more about this... I guess the advantage of the on-the-fly heap would be that you may never have to calculate the ordering for some chunks in the middle of the mempool (if they get rbf'ed, or conflicted by a block tx, or if they expire, or if they get cpfp'ed and thus brought to the front of the mempool, rather than just waiting until the best block feerate drops to their level, or you stop and restart your bitcoind before they get mined). Probably few enough txs hit those special cases to be much of a saving overall though

Another thought: consider the "virtual Cluster class" idea, but instead of special casing each standalone tx into a custom single-tx-Cluster, you could have a single cluster for all the standalone txs. That would then just order those txs by feerate, but it would mean that instead of having to maintain a separate chunk/cluster data structure for each of those txs, you only maintain a single cluster data structure that covers all of them. That approach messes up the model for diagram comparisons that this PR introduces (because suddenly the diagram comparisons pull in unrelated singleton txs whenever any one of them is touched), but maybe that could be fixed reasonably.

ajtowns replied:
> A downside here is that the "linearization" for this fat singletons cluster would be O(n log n) to update on every operation involving a singleton, which may be substantial if there are 1000s of singletons?

I don't think that's true? You'd just store the linearisation as part of the special cluster implementation as a std::set by feerate order, for O(lg n) modification and O(n) reporting? I guess it would be better to call that class something other than "cluster", since you'd probably want `GetCluster(singleton)` to just return the singleton, and `CountDistinctClusters({singleton1, singleton2})` to return 2, etc.

> A variation on this idea though: add a new QualityLevel::SINGLETON, and allow many singleton clusters at the same time, each individually limited to 64 (or whatever) singleton transactions. This might not even need the virtual class idea.

Hmm, maybe -- you could find a new child tx would spend 63 txs from 63 different singleton clusters, but I don't know if there's anything about that that would be worse than the number of clusters you could impact by conflicting with pre-existing txs; and 63*64 is only 4032 which doesn't sound very scary... Big mempools seem like they might contain up to 175k txs (300MB limit; or 350k txs unlimited), if they were all singletons, dividing by 64 gives just 2700 (or 5500) clusters, which seems nice. (I guess you'd need something to keep track of which existing singleton clusters aren't full and can be added to)

instagibbs commented on src/txgraph.h:
```diff
@@ -167,6 +169,28 @@ class TxGraph
+    };
+
+    /** Construct a block builder, drawing chunks in order, from the main graph, which cannot be
+     *  oversized. While the returned object exists, no mutators on the main graph are allowed. */
```
Sorry, that the builder *doesn't* outlive the graph. I was testing to see what happens when it *does* outlive it, and it's not crashing currently.

instagibbs reviewed (approved):
ACK d8e4b92fb4aea16e18285d25c4d5bba3b3b51c98

non-blocking comments

`git range-diff master 0630995fee22990402771547be1480b8706c76ce d8e4b92fb4aea16e18285d25c4d5bba3b3b51c98`

sipa reviewed (commented):
Rebased on top of #32177 and #32191.

instagibbs commented on src/txgraph.h:
```diff
@@ -167,6 +169,28 @@ class TxGraph
+        /** Mark the current chunk as included, and progress to the next one. */
+        virtual void Include() noexcept = 0;
+        /** Mark the current chunk as skipped, and progress to the next one. Further chunks from
+         *  the same cluster as the current one will not be reported anymore. */
```
If the top `GetCurrentChunk` returns nullopt, it hits an early `break`. so it doesn't get to hit a no-op Skip/Include?

instagibbs replied:
looks right, and instantly crashes when I add an Assume in the right spot

instagibbs commented:
reACK 27a0c93abb7e70b93214eb857e2046f848139e68

addressed comments and put the choice to just get next chunk again under fuzzer input directly.

instagibbs commented:
reACK d64bd31f9b6903f0a335a78b519d0b52e84cca1b

Simple rebase on top of stability fixes.

`git range-diff master 27a0c93abb7e70b93214eb857e2046f848139e68 d64bd31f9b6903f0a335a78b519d0b52e84cca1b`

instagibbs commented:
ACK a0becaaa9c7390bc80d5864f7fffb32e21502a50

Added assertion, cleaned up comment, and rebase on master

`git range-diff master d64bd31f9b6903f0a335a78b519d0b52e84cca1b a0becaaa9c7390bc80d5864f7fffb32e21502a50`

ismaelsadeeq commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -639,6 +717,62 @@ FUZZ_TARGET(txgraph)
+            // been omitted).
+            assert(missing_main_cmp == missing_stage_cmp);
+
+            // The missing part must include at least all transactions in staging which have not been
+            // modified, or been in a cluster together with modified transactions, since they were
+            // copied from main. Note that due to the reordering of removals w.r.t. dependency
+            // additions, it is possible that the real implementation found more unaffected things.
```
This explanation clarified the comment, thanks.

ismaelsadeeq reviewed (commented):
Code review ACK a0becaaa9c7390bc80d5864f7fffb32e21502a50

Comments are just nits and some questions.

instagibbs commented on src/test/fuzz/txgraph.cpp:
```diff
@@ -639,6 +717,62 @@ FUZZ_TARGET(txgraph)
+        auto main_implied_diagram = ChunkLinearization(sims[0].graph, vec1);
+        assert(CompareChunks(main_real_diagram, main_implied_diagram) == 0);
+
+        if (sims.size() >= 2 && !sims[1].IsOversized()) {
```
this was already checked for earlier in the first if block

instagibbs reviewed (approved):
reACK 7208ec2f1f0b9d72e1b4248dae9adbcc5ab625f7

`git range-diff master a0becaaa9c7390bc80d5864f7fffb32e21502a50 7208ec2f1f0b9d72e1b4248dae9adbcc5ab625f7`

ismaelsadeeq commented:
reACK https://github.com/bitcoin/bitcoin/commit/7208ec2f1f0b9d72e1b4248dae9adbcc5ab625f7

monlovesmango reviewed (commented):
ACK 7208ec2f1f0b9d72e1b4248dae9adbcc5ab625f7

All my comments are mostly just questions.

Sorry it took a while to do my review, just wanted to make sure I actually felt like I had a good understanding of all the changes (which took a while).

Unrelated to this PR and more of an FYI, in src/test/fuzz/txgraph.cpp `pick_fn` I think that:
```c 
size_t choices = tx_count[0] + sims[0].removed.size() + 1;
```
can just be:
```c
size_t choices = tx_count[0] + sims[0].removed.size();
```
and: 
```c
auto choice = provider.ConsumeIntegralInRange<size_t>(0, choices - 1);
```
can just be:
```c
auto choice = provider.ConsumeIntegralInRange<size_t>(0, choices);
```

glozow reviewed (commented):
ACK 7208ec2f1f0

Light code review + tested by [implementing](https://github.com/glozow/bitcoin/commits/2025-04-package-validation/) a package linearizer + chunker on top of `TxGraph` and running #26711 unit tests. It is linearizing and chunking the transactions as I expect, which is great (it still has trouble with cases like RBF, chunks not fitting, and skipping non-descendants, but definitely out of scope for this PR).

sipa committed:
txgraph: abstract out transaction ordering (refactor)

sipa reviewed (commented):
> Unrelated to this PR and more of an FYI, in src/test/fuzz/txgraph.cpp pick_fn I think that:

Sure, that would be correct, but very confusing, I think!

The number of possible choices is `tx_count[0] + sims[0].removed.size() + 1`, and this value is stored in the variable `choices`, representing the number of choices.

A number is then generated in the range between `0` and `choices - 1`, inclusive.

Of course, operationally, the `+ 1` and subsequent `- 1` are superfluous, but then the variable would probably need to be renamed to `choices_minus_one`, or `max_valid_choice`.

monlovesmango commented on src/txgraph.h:
```diff
@@ -162,6 +164,39 @@ class TxGraph
+    public:
+        /** Support safe inheritance. */
+        virtual ~BlockBuilder() = default;
+        /** Get the chunk that is currently suggested to be included, plus its feerate, if any. */
```
Sorry yes you are correct. I misread the usage of `m_cur_cluster` (and then proceeded to also incorrectly fuzz test when trying to confirm my understanding).

monlovesmango commented on src/txgraph.cpp:
```diff
@@ -498,6 +564,66 @@ const TxGraphImpl::ClusterSet& TxGraphImpl::GetClusterSet(int level) const noexc
+void TxGraphImpl::ClearChunkData(Entry& entry) noexcept
+{
+    if (entry.m_main_chunkindex_iterator != m_main_chunkindex.end()) {
+        Assume(m_main_chunkindex_observers == 0);
```
Mostly just trying to get an understanding of your thought process on when an Assume() is appropriate. That is good context to know about Assume() usage.

I expected it in the higher-level functions `AddTransaction`, `RemoveTransaction`, `AddDependency`, `SetTransactionFee`, and `CommitStaging` as these all can change linearization of main graph and did indeed find it in these.

I didn't expect it in `ClearChunkData` and `UnlinkRef`, but thought maybe it was there to be confident that we don't remove data that `BlockBuilderImpl` might need/reference. However if this was the correct rationale I would have also expected `ClearChunkData` to have an `Assume(m_main_chunkindex_observers == 0)`. So was just wondering what the rationale was if there was one.

monlovesmango commented:
> The number of possible choices is `tx_count[0] + sims[0].removed.size() + 1`, and this value is stored in the variable `choices`, representing the number of choices.

Interesting, I guess my interpretation is different and don't understand where the +1 comes from. I would have thought the number of choices is simply the tx count plus the removed tx count.

> Sure, that would be correct, but very confusing, I think!

Its funny, I only noticed this because I found the current assignment very confusing hahah.

Anyway, I think this can be ignored then.

monlovesmango commented:
reACK 8673e8f01917b48a5f5476792f759f44ea49d5a5

glozow commented:
reACK 8673e8f0191

glozow reviewed (commented):
reACK 8673e8f0191
